src/cockpit/channel.py

```diff
@@ -397,3 +397,35 @@ def do_data(self, data):
             data = bytes(data)
 
         self.receive_queue.put_nowait(data)
+
+
+class GeneratorChannel(Channel):
+    """A trivial Channel subclass for sending data from a generator with flow control.
+
+    Calls the .get_data() generator with the options from the open message and
+    sends the data which it yields.  If the generator returns a value it will
+    be used for the close message.
+    """
+    __generator: Generator[bytes, int, Optional[Dict[str, object]]]
+
+    def get_data(self, size_hint: int, options: dict[str, object]) -> Generator[bytes, int, Optional[Dict[str, object]]]:
+        raise NotImplementedError
+
+    def do_open(self, options: Dict[str, object]) -> None:
+        self.__generator = self.get_data(Channel.CHANNEL_FLOW_WINDOW, options)
+        self.iterate(None)  # first Generator.send() call needs to be with None
+
+    def iterate(self, n_bytes: Optional[int]) -> None:
+        try:
+            while n_bytes is None or n_bytes > 0:
```

Comment:
This could infinite loop if the generator keeps yielding empty blocks.  That would happen is you tried to fsread `/dev/zero`, for example, since we drop `'\0'` characters.

Reply:
We drop `\0`? That feels like quite a grave bug in itself? Why is that, python bytearrays should be fine with that?

```
❱❱❱ python3 -c 's = b"a\0b"; print(len(s), s)'
3 b'a\x00b'
```

(unrelated of course, but this should be tracked then)